nifi-nar-bundles/nifi-standard-services/nifi-hbase_1_1_2-client-service-bundle/nifi-hbase_1_1_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_1_1_2_ClientMapCacheService.java

```diff
@@ -0,0 +1,224 @@
+    private HBaseClientService hBaseClientService;
+
+    private String hBaseColumnFamily;
+    private byte[] hBaseColumnFamilyBytes;
+
+    private String hBaseColumnQualifier;
+    private byte[] hBaseColumnQualifierBytes;
+
+    @OnEnabled
+    public void onConfigured(final ConfigurationContext context) throws InitializationException{
+       hBaseCacheTableName  = context.getProperty(HBASE_CACHE_TABLE_NAME).getValue();
+       hBaseClientService   = context.getProperty(HBASE_CLIENT_SERVICE).asControllerService(HBaseClientService.class);
+       hBaseColumnFamily    = context.getProperty(HBASE_COLUMN_FAMILY).getValue();
+       hBaseColumnQualifier = context.getProperty(HBASE_COLUMN_QUALIFIER).getValue();
+
+       hBaseColumnFamilyBytes    = hBaseColumnFamily.getBytes(StandardCharsets.UTF_8);
+       hBaseColumnQualifierBytes = hBaseColumnQualifier.getBytes(StandardCharsets.UTF_8);
+    }
+
+    private <T> byte[] serialize(final T value, final Serializer<T> serializer) throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        serializer.serialize(value, baos);
+        return baos.toByteArray();
+    }
+    private <T> T deserialize(final byte[] value, final Deserializer<T> deserializer) throws IOException {
+        return deserializer.deserialize(value);
+    }
+
+
+    @Override
+    public <K, V> boolean putIfAbsent(final K key, final V value, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) throws IOException {
+      if (containsKey(key, keySerializer)) {
+        put(key, value, keySerializer, valueSerializer);
+        return true;
+      } else return false;
+    }
+
+    @Override
+    public <K, V> void put(final K key, final V value, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) throws IOException {
+
+        List<PutColumn> putColumns = new ArrayList<PutColumn>(1);
+        final byte[] rowIdBytes = serialize(key, keySerializer);
+        final byte[] valueBytes = serialize(value, valueSerializer);
+
+        final PutColumn putColumn = new PutColumn(hBaseColumnFamilyBytes, hBaseColumnQualifierBytes, valueBytes);
+        putColumns.add(putColumn);
+
+        hBaseClientService.put(hBaseCacheTableName, rowIdBytes, putColumns);
+    }
+
+    @Override
+    public <K> boolean containsKey(final K key, final Serializer<K> keySerializer) throws IOException {
+      final byte[] rowIdBytes = serialize(key, keySerializer);
+      final HBaseRowHandler handler = new HBaseRowHandler();
+
+      final List<Column> columnsList = new ArrayList<Column>(0);
+
+      hBaseClientService.scan(hBaseCacheTableName, rowIdBytes, rowIdBytes, columnsList, handler);
+      return (handler.numRows() > 0);
+    }
+
+    @Override
+    public <K, V> V getAndPutIfAbsent(final K key, final V value, final Serializer<K> keySerializer, final Serializer<V> valueSerializer, final Deserializer<V> valueDeserializer) throws IOException {
```

Comment:
For putIfAbsent and getAndPutIfAbsent, they currently require two client calls (a get + conditional write). This could be problematic in that you can't ensure that after doing the get that something else hasn't modified the value before you do the write, this could happen for example in a cluster where two nodes go to do a putIfAbsent at the same time with the same key. 

It looks like HBase has some atomic checkAndMutate operations:
https://hbase.apache.org/apidocs/org/apache/hadoop/hbase/client/Table.html

I think we should see if we can use these in place of the two calls, although admittedly I have not used them myself.